src/HtmlUtils.js

```diff
@@ -142,28 +136,40 @@ module.exports = {
 
         var isHtml = (content.format === "org.matrix.custom.html");
 
-        var safeBody;
+        var safeBody, body;
         if (isHtml) {
+            // XXX: We sanitize the HTML whilst also highlighting its text nodes, to avoid accidentally trying
+            // to highlight HTML tags themselves.  However, this does mean that we don't highlight textnodes which
+            // are interrupted by HTML tags (not that we did before) - e.g. foo<span/>bar won't get highlighted
+            // by an attempt to search for 'foobar'.  Then again, the search query probably wouldn't work either
+            if (highlights && highlights.length > 0) {
+                var highlighter = new Highlighter(isHtml, "mx_EventTile_searchHighlight", opts.onHighlightClick);
+                var safeHighlights = highlights.map(function(highlight) {
+                    return sanitizeHtml(highlight, sanitizeHtmlParams);
+                });
+                // XXX: hacky bodge to temporarily apply a textFilter to the sanitizeHtmlParams structure.
+                sanitizeHtmlParams.textFilter = function(safeText) {
+                    return highlighter.applyHighlights(safeText, safeHighlights).map(function(span) {
+                        // XXX: rather clunky conversion from the react nodes returned by applyHighlights
+                        // (which need to be nodes for the non-html highlighting case), to convert them
```

Comment:
the non-html case takes nodes and returns them straight to the render method, where react directly consumes the spans and textnodes (escaping the textnodes as needed).

the html-case takes nodes but has to convert them into an HTML string as sanitize-html's textFilter operates strictly on strings-of-HTML rather than react nodes.  So the conversion and escaping is done by sanitize-html and then added into react via dangerouslySetInnerHTML.